
#Бронирование отелей

В этом мини-проекте проанализированы данные о бронировании отелей. Чуть ниже предоставлено описание этих данных и то, как они выглядят.

In [2]:
import pandas as pd

In [3]:
bookings = pd.read_csv('https://stepik.org/media/attachments/lesson/360344/bookings.csv', sep = ';')
bookings.head(7)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [4]:
bookings.shape

(119390, 21)

In [5]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [6]:
# Переименуем колонки

def replace_space_with_star(name):
    new_name = name.replace(' ', '_').lower()
    return new_name

bookings = bookings.rename(columns=replace_space_with_star)

###ТОП - 5 стран, пользователи из которых совершили наибольшее число успешных бронирований


In [13]:
bookings[(bookings['is_canceled'] == False)].country.value_counts().head(5)

country
PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: count, dtype: int64

##Среднее количество ночей, на которое бронируют отели разных типов

In [20]:
bookings.groupby('hotel').agg({"stays_total_nights" : "mean"}).round(2)

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

##Количество несовпадений номера, полученного клиентом и изначально забронированного номера

In [34]:
bookings.query('assigned_room_type == reserved_room_type').agg({'reserved_room_type': 'count'})

reserved_room_type    104473
dtype: int64

##Проанализируем запланированные даты прибытия

Согласно данным ниже, самым популярным месяцем в 2016 году был октябрь, а в 2017 - май.

In [41]:
bookings[(bookings['arrival_date_year']==2016)].value_counts('arrival_date_month')

arrival_date_month
October      6203
May          5478
April        5428
September    5394
June         5292
August       5063
March        4824
July         4572
November     4454
February     3891
December     3860
January      2248
Name: count, dtype: int64

In [42]:
bookings[(bookings['arrival_date_year']==2017)].value_counts('arrival_date_month')

arrival_date_month
May         6313
April       5661
June        5647
July        5313
March       4970
August      4925
February    4177
January     3681
Name: count, dtype: int64

##Число отмен броней по месяцам каждого года для типа отеля "City Hotel"

Согласно данным ниже, в 2015 году больше всего отмен произошло в Сентябре, в 2016 - в Октябре, и в 2017 - в Мае

In [52]:
bookings[((bookings['hotel']=='City Hotel') & (bookings['is_canceled']==True))].groupby(['arrival_date_year'])['arrival_date_month'].value_counts()

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1808
                   July                  1324
                   March                 1278
                   August                1123
                   January               1044
                   February               971
Name: count, dtype: int64

##Среднее количество колонок adults, children и babies.

Из них наибольшее среднее значение имеет adults

In [53]:
bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

##Среднее значение переменной total_kids для разных типов отелей

total_kids - общее количество children и babies.

Для Resort Hotel это значение больше

In [54]:
bookings['total_kids'] = bookings['children'] + bookings['babies']

In [57]:
bookings[['hotel', 'total_kids']].groupby('hotel').agg({'total_kids' : 'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

## Churn rate

Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе.

Churn rate (отток, коэффициент оттока) – это процент подписчиков (например, на push-уведомления от сайта), которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах.

В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов!

Создадим переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False.


In [60]:
bookings['has_kids'] = (bookings['total_kids']>0)

In [79]:
#Churn rate для клиентов с детьми
round(bookings[((bookings['has_kids']==True) & (bookings['is_canceled']==True))].shape[0]/bookings[(bookings['has_kids']==True)].shape[0]*100,2)

34.92

In [80]:
#Churn rate для клиентов без детей
round(bookings[((bookings['has_kids']==False) & (bookings['is_canceled']==True))].shape[0]/bookings[(bookings['has_kids']==False)].shape[0]*100,2)

37.22

Вывод: Можно заметить, что бОльший отток клиентов произошел в группе, запланировавших отпуск без детей.

Так метрика Churn rate составила 34.92 в группе клиентов с детьми, и 37.22 в группе без детей.